### PR TITLE
Implementing electron-compile for Yeoman App

### PR DESCRIPTION
**Pull request by mfunkie**
There's still some future work to be done here.

See: https://github.com/paulcbetts/electron-compile#how-can-i-precompile-my-code-for-release-time

The idea for release time would be to pre-compile everything on the command line in the build script, and then exclude the compilers folder in node modules in the ignore part of packagerOptions in `scripts/build.js` so that we don't ship all the compilers with the App.  In doing this, we would need to either set some environment variable in our app or some way to communicate to our `main.js` file that we're running in production so that instead of `require('electron-compile').initWithOptions`, we would `require('electron-compile').initForProduction`.

This gets us fairly far though and enables ES6 and beyond development within the app at the temporary (technical debt) cost of having 50 extra MB added to the application size.  I've updated `src/browser/application.js` to be an ES6 class as both an example and a test for the ES6 usage.

### CODE REVIEW

**Comment by sindresorhus** [comment]
// @paulcbetts

**Comment by paulcbetts** [comment]
Cool!

> In doing this, we would need to either set some environment variable in our app or some way to communicate to our main.js file that we're running in production so that instead of require('electron-compile').initWithOptions, we would require('electron-compile').initForProduction.

So, I'm not sure how your build process works, but effectively in development mode, we don't generate a cache at all (we let electron-compile use the default cache which is in $TMPDIR). Then, we just use the presence of the cache directory tell us whether to initForProduction (this is only 1% more complicated for our code because we have this concept of 'devMode'):

```js
var fs = require('fs');
var path = require('path');
var _ = require('lodash');

// ROOT/src/browser/../../cache
var cachePath = path.join(__dirname, '..', '..', 'cache');
var devMode = _.find(process.argv, function(x) { return x === '-r' });

if (fs.statSyncNoException(cachePath) && !devMode) {
  require('electron-compile').initForProduction(cachePath);
} else {
  require('electron-compile').init();
}

require('./main');
```

**Comment by mfunkie** [comment]
@paulcbetts 

I've made some updates to the code locally that mirror what you're doing, but I'm not sure how I would tell the Production build to ignore certain files.  In the production build I get a white screen that says

```
"Failed to compile /Users/mfunk/Documents/yeoman-app/build/darwin/x64/Yeoman.app/Contents/Resources/app.asar/static/index.html: Asked to compile /Users/mfunk/Documents/yeoman-app/build/darwin/x64/Yeoman.app/Contents/Resources/app.asar/static/index.html in production!
Error: Asked to compile /Users/mfunk/Documents/yeoman-app/build/darwin/x64/Yeoman.app/Contents/Resources/app.asar/static/index.html in production!
    at ReadOnlyCompiler.compile (/Users/mfunk/Documents/yeoman-app/build/darwin/x64/Yeoman.app/Contents/Resources/app.asar/node_modules/electron-compile/lib/read-only-compiler.js:41:13)
    at ReadOnlyCompiler.loadFile (/Users/mfunk/Documents/yeoman-app/build/darwin/x64/Yeoman.app/Contents/Resources/app.asar/node_modules/electron-compile/node_modules/electron-compile-cache/lib/compile-cache.js:217:19)
    at Function."
```

**Comment by mfunkie** [comment]
I've added my current code to the PR that will still has the index.html error, maybe someone else can take a look and help me figure out what I'm doing wrong?

**Comment by paulcbetts** [comment]
@mfunkie I'll have a look in a bit. 

To be honest though, you probably don't need a `static` folder at all if you didn't want one. `electron-compile` is kind of a paradigm change vs. what you're used to doing with web dev, the way to think about it is, pretend as if Electron / Chromium just natively understands ES2015/LESS/etc, you don't need to jump through any special build hoops other than running `electron-compile` (in fact, having production dirs have different names than the source ones will break stuff).

**Comment by mfunkie** [comment]
I think the main problem area is how the browser-window module is loading the url that is index.js

https://github.com/yeoman/yeoman-app/blob/master/src/browser/appwindow.js#L42

Is there a sort of blessed way of opening the main application window in a post `electron-compile` world?

**Comment by paulcbetts** [comment]
> Is there a sort of blessed way of opening the main application window in a post electron-compile world?

There shouldn't be any special way in particular, it should just work

**Comment by paulcbetts** [comment]
@mfunkie I'm not sure why I didn't catch this earlier, but this is a dumb bug in electron-compile's protocol hook. If you bump to electron-compile@0.7.1, yeoman-app is now working

**Comment by mfunkie** [comment]
Thanks for fixing the bug, @paulcbetts !

@sindresorhus I've tested this working locally and in a built version.  Let me know if there's anything else I can do to make this PR happen :)

**Comment by stefanbuck** [comment]
@mfunkie thanks for all the effort :+1: :+1:  Will take a closer look on it this evening. 

Also big thanks @paulcbetts for the instant help :exclamation:

**Comment by ruyadorno** [comment]
@mfunkie looking really good :+1: thanks again

File: src/browser/application.js [hunk 1]
```diff
@@ -1,170 +1,167 @@
...
-}
+import app              from 'app';
+import ipc              from 'ipc';
+import path             from 'path';
+import shell            from 'shell';
+import BrowserWindow    from 'browser-window';
+import { EventEmitter } from 'events';
+import _                from 'underscore-plus';
+import AppMenu          from './appmenu';
+import AppWindow        from './appwindow';
```
Comment (stefanbuck) [inline src/browser/application.js]: Only one smallish thing left. Could you please remove the extra spaces to be in line with the repo style. Thank you

Comment (mfunkie) [inline src/browser/application.js]: Damn, thought I could slip that style preference past you ;)

Comment (ruyadorno) [inline src/browser/application.js]: hahahah, nice try @mfunkie :smirk:

**Comment by mfunkie** [comment]
Woot! Now comes the fun stuff...

**Comment by ruyadorno** [comment]
yay!!! :confetti_ball: :tada:

**Comment by mfunkie** [comment]
We'll make this choose your own adventure, as I can only help with one thing at a time.

Do you want:
* Convert all files to ES6 usage (potential fewer lines of code/cleaner code)
* Update Flux implementation to actually be flux (Store shouldn't emit specific actions and should hold the data themselves, would really clean up the logic)
* React Router Implementation

**Comment by stefanbuck** [comment]
I'm up for "Update Flux implementation"

cc @ruyadorno

**Comment by bitshadow** [comment]
I started getting https://github.com/atom/atom/issues/6632 in Maverick 10.9.4. After checking the logs I found this commit is causing this error for me. I am still having this issue. Any idea?

**Comment by paulcbetts** [comment]
@bitshadow What's the console output when that happens

**Comment by bitshadow** [comment]
@paulcbetts 
```
> yeoman-app@0.0.0 start /Users/jignesh.kakadiya/projects/self/test/yeoman-app
> electron . -r

App load time: 975ms
[358:0730/174451:INFO:CONSOLE(25)] "Error: UnknownSystemError: Unknown system error, uv_pipe_open
    at Error (native)
    at new Socket (net.js:126:18)
    at createWritableStdioStream (node.js:600:18)
    at process.stdout (node.js:626:16)
    at /Users/jignesh.kakadiya/projects/self/test/yeoman-app/node_modules/electron-compile/node_modules/electron-compilers/node_modules/babel-core/node_modules/chalk/node_modules/supports-color/index.js:29:13
    at Object.<anonymous> (/Users/jignesh.kakadiya/projects/self/test/yeoman-app/node_modules/electron-compile/node_modules/electron-compilers/node_modules/babel-core/node_modules/chalk/node_modules/supports-color/index.js:50:3)
    at Module._compile (module.js:426:26)
    at BabelCompiler.loadFile (/Users/jignesh.kakadiya/projects/self/test/yeoman-app/node_modules/electron-compile/node_modules/electron-compile-cache/lib/compile-cache.js:199:23)
    at Object.value [as .js] (/Users/jignesh.kakadiya/projects/self/test/yeoman-app/node_modules/electron-compile/node_modules/electron-compile-cache/lib/compile-cache.js:240:26)
    at Module.load (module.js:351:32)", source: file:///Users/jignesh.kakadiya/projects/self/test/yeoman-app/static/index.js (25)
YO-Process { action: 'generator:init', args: [] }
```

**Comment by paulcbetts** [comment]
@bitshadow This is because you're trying to write to stdout, when it's closed sometimes on OS X. Unfortunately you probably need to patch chalk or not load it at all

**Pull request merged**